server/src/client/session.rs

```diff
@@ -208,3 +104,663 @@ impl Handler<SendMessage<ServerMessage>> for ClientWsSession {
+    }
+
+    fn change_password(
+        &mut self,
+        old_password: String,
+        new_password: String,
+        user_id: UserId,
+        request_id: RequestId,
+        ctx: &mut WebsocketContext<Self>,
+    ) {
+        if !auth::valid_password(&new_password, self.config.get_ref()) {
+            return ctx.binary(ServerMessage::Response {
+                response: RequestResponse::Error(ServerError::InvalidPassword),
+                request_id,
+            });
+        }
+
+        let fut = self
+            .verify_user_id_password(user_id, old_password)
+            .and_then(|res| match res {
+                Ok(_) => Either::A(auth::hash(new_password).map(|ok| Ok(ok))),
+                Err(error) => Either::B(future::ok(Err(error))),
+            })
+            .into_actor(self)
+            .and_then(move |res, act, _ctx| {
+                let fut = match res {
+                    Ok((new_password_hash, hash_version)) => {
+                        let fut = act
+                            .database_server
+                            .send(ChangePassword {
+                                user_id,
+                                new_password_hash,
+                                hash_version,
+                            })
+                            .map(move |res| res.map(|_| ()))
+                            .map(|res| match res {
+                                Ok(_) => RequestResponse::success(),
+                                Err(l337::Error::Internal(e)) => {
+                                    eprintln!("Database connection pooling error: {:?}", e);
+                                    RequestResponse::Error(ServerError::Internal)
+                                }
+                                Err(l337::Error::External(sql_error)) => {
+                                    eprintln!("Database error: {:?}", sql_error);
+                                    RequestResponse::Error(ServerError::Internal)
+                                }
+                            });
+                        Either::A(fut)
+                    }
+                    Err(e) => Either::B(future::ok(RequestResponse::Error(e))),
+                };
+
+                fut.into_actor(act)
+            })
+            .and_then(move |res, act, _ctx| match res {
+                RequestResponse::Success(success) => fut::Either::A(
+                    act.client_server
+                        .send(LogoutUserSessions { user_id })
+                        .map(|_| RequestResponse::Success(success))
+                        .into_actor(act),
+                ),
+                response => fut::Either::B(fut::ok(response)),
+            });
+
+        self.respond(fut, request_id, ctx)
+    }
+
+    fn verify_user_id_password(
+        &mut self,
+        user_id: UserId,
+        password: String,
+    ) -> impl Future<Item = Result<(), ServerError>, Error = MailboxError> {
+        self.database_server
```

Comment:
Interaction with the database server could maybe be extracted to a function for requests specifically. Like it'd map the database errors to internal server error.

Reply:
Huh? What do you mean?

Reply:
There's a lot of repetition in these similar functions that map the error from database to an internal server error type -- I'm proposing a function that would make a request to the database via send but map the error too such that it doesn't have to be repeated